### ISSUE TITLE
selfcal sofia crashes with input mask

### ISSUE THREAD

**Issue by healytwin1** [caracal-pipeline/caracal#1140]
I get the following error when I specify an input mask in the self cal img_sofia settings:

```
# --- SoFiA 1.3.2: Running source finder ---------------------------------------
#     Elapsed time: 00:00:02.80 h
# 
# Running S+C filter
#     Estimating rms on subcube (x,y,z zoom = 1,1,1) ...
#     Estimating rms on subcube sampling every 2 voxels ...
#     ... Subcube shape is (1, 2160, 2160) ...
#     ... mad rms = 1.00e+00 (data units)
# 
#     Elapsed time: 00:00:02.90 h
#     Filter 0.0 0.0 0 b ...
#     Generating noise-scaled data cube:
#       Selecting local noise measurement method.
#       Applying median absolute deviation to all pixels.
#       Using grid size of [35 pix , 1 chan ]
#       and window size of [71 pix , 1 chan ].
#       Interpolating in between grid points (linear).
#     Noise-scaled data cube generated.
# 
#     Estimating rms on subcube (x,y,z zoom = 1,1,1) ...
#     Estimating rms on subcube sampling every 2 voxels ...
#     ... Subcube shape is (1, 2160, 2160) ...
#     ... mad rms = 9.99e-01 (data units)
#     Applying +/- 10.0 sigma detection threshold
# Traceback (most recent call last):
#   File "/sofia/sofia_pipeline.py", line 370, in <module>
#     pyfind.SCfinder_mem(np_Cube, mask, dict_Header, t0, **Parameters["SCfind"])
#   File "/sofia/build/lib.linux-x86_64-2.7/sofia/pyfind.py", line 79, in SCfinder_mem
#     mask |= (np.absolute(cube_smooth) >= threshold * rms_smooth)
# TypeError: ufunc 'bitwise_or' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
# Traceback (most recent call last):
#   File "/stimela_mount/code/run.py", line 61, in <module>
#     subprocess.check_call(shlex.split(_runc))
#   File "/usr/lib/python2.7/subprocess.py", line 190, in check_call
#     raise CalledProcessError(retcode, cmd)
# subprocess.CalledProcessError: Command '['sofia_pipeline.py', 'sofia_parameters.par']' returned non-zero exit status 1
2020-05-25 12:12:37 STIMELA.selfcal INFO: Completed jobs : ['version-HighZ_E1_v2_selfcal_before-ms0', 'image-field0-iter0', 'make-sofia_mask-field0-iter0-import-image', 'make-sofia_mask-field0-iter0-import-mask', 'make-sofia_mask-field0-iter0-regrid', 'make-sofia_mask-field0-iter0-export-mosaic', 'make-sofia_mask-field0-iter0-export-mask']
```

I created the mask by running sofia on my image and cleaning that up (values are 0 or 1).

**Comment by paoloserra** [comment]
Could you please check whether https://github.com/SoFiA-Admin/SoFiA/issues/159 is relevant in this case? (Though it should have been fixed in SoFiA.)

**Comment by paoloserra** [comment]
Also, can you confirm that you want to add additional detections to the mask you have made? If, on t he contrary, you just want to use this mask without adding anything to it, you should use different settings.

**Comment by healytwin1** [comment]
> Could you please check whether [SoFiA-Admin/SoFiA#159](https://github.com/SoFiA-Admin/SoFiA/issues/159) is relevant in this case? (Though it should have been fixed in SoFiA.)

It seems to me like that this issue is relevant. I have deleted the relevant B* keywords and restarted, so will see how that goes.


> Also, can you confirm that you want to add additional detections to the mask you have made? If, on t he contrary, you just want to use this mask without adding anything to it, you should use different settings.

Yes, I am wanting to add to the mask.

**Comment by healytwin1** [comment]
okay, can confirm that I get the same error even with the B* keywords removed.

**Comment by paoloserra** [comment]
Would you be able to try this with SoFiA 1.3.2 outside CARACal? As I said, in principle this was fixed in that SoFiA release, which is the one then included in Stimela. I will also try to reproduce your error.

**Comment by paoloserra** [comment]
@healytwin1 a quick update to say that I can reproduce your error. I'll continue looking into this.

**Comment by healytwin1** [comment]
> Would you be able to try this with SoFiA 1.3.2 outside CARACal? 

I'm not going to have the chance in the next couple of days, need to get some other stuff finished

**Comment by paoloserra** [comment]
This is a CARACal bug.

SoFiA always writes integer masks to disc. https://github.com/SoFiA-Admin/SoFiA/issues/159 had to do with integer masks being incorrectly imported as float-type masks by SoFiA. That issue was fixed in SoFiA and does not affect CARACal.

Your mask, too, used to be of integer type. It should have  caused no problems.

However, I've just found out that when you give an input mask to CARACal the `selfcal` worker sneakily regrids it (whether that was necessary or -- as in  your case -- not). In the process, the mask is converted from its original integer type into float (and overwritten!!!). See lines:

https://github.com/caracal-pipeline/caracal/blob/2caf475b9c50db1006da643e6d0613f7bbf20f57/caracal/workers/selfcal_worker.py#L684-L748

Shame on us for allowing this to happen.

The reason for regridding is that it would allow you to give an input mask made from an image on a different WCS grid. That was the plan anyway...

I'll include this in our list of bug fixes for the next CARACal minor release. Apologies for the trouble.
